src/lib/spacefinder/liveblog-right-adverts.ts

```diff
@@ -0,0 +1,75 @@
+import type { Advert } from 'lib/dfp/Advert';
+import { createAdvert } from 'lib/dfp/create-advert';
+import { dfpEnv } from 'lib/dfp/dfp-env';
+import { displayAds } from 'lib/dfp/display-ads';
+import { displayLazyAds } from 'lib/dfp/display-lazy-ads';
+import { queueAdvert } from 'lib/dfp/queue-advert';
+import { isInEagerPrebidVariant } from 'lib/experiments/eager-prebid-check';
+import { requestBidsForAds } from '../header-bidding/request-bids';
+
+type AdsInsertedCustomEventDetail = {
+	numAdsToInsert: number;
+	fromIndex: number;
+};
+
+type AdsInsertedCustomEvent = {
+	detail: AdsInsertedCustomEventDetail;
+};
+
+const isCustomEvent = (event: Event): event is CustomEvent => {
+	return 'detail' in event;
+};
+
+// TODO: There's lots of overlap with fill-advert-slots.ts in this function. If multiple right
+// ad slots on liveblogs is what we go forward with, then refactor this to use common functions
+const fillAdvertSlots = async (numAdsToInsert: number, fromIndex: number) => {
+	// Create an array with the indexes of adverts that need to be created
+	// This is to exclude adverts that have already been loaded
+	const newAdvertIndexes = [...Array(numAdsToInsert).keys()].map(
+		(i) => i + fromIndex,
+	);
+
+	const adverts = [
+		...document.querySelectorAll<HTMLElement>('.ad-slot--liveblog-right'),
+	]
+		.filter(({ id }) =>
+			// exclude ad slots that have already been filled
+			newAdvertIndexes.includes(Number(id[id.length - 1])),
+		)
+		.map((adSlot) => createAdvert(adSlot))
+		.filter((advert): advert is Advert => advert !== null);
+
+	if (!adverts.length) return;
+
+	const currentLength = dfpEnv.adverts.length;
+	dfpEnv.adverts = dfpEnv.adverts.concat(adverts);
+	adverts.forEach((advert, index) => {
+		dfpEnv.advertIds[advert.id] = currentLength + index;
+	});
+
+	if (isInEagerPrebidVariant()) {
+		// Request bids for all server rendered adverts
+		await requestBidsForAds(adverts);
+	}
+
+	adverts.forEach(queueAdvert);
+	if (dfpEnv.shouldLazyLoad()) {
```

Comment:
I think you can avoid duplicating the code that directly manipulates `dfpEnv.adverts` and calls `displayAds` here by calling [`addSlot`](https://github.com/guardian/commercial-core/blob/d0ec41f494be36556ad1a25a889556e5b3cce921/src/lib/dfp/add-slot.ts#L20). I think this is the pattern taken by `article-body-adverts.ts` and `liveblog-adverts.ts`.

EDIT: It'll handle the call to `createAdvert` as well.

Reply:
Good idea, will update. 

It does feel like `addSlot` is a weird name. The function takes a slot and fills that slot with an advert, but `addSlots` sounds like we're adding a slot to the page. What do you think about changing this name to `fill-advert-slot-dynamic` or `fill-dynamic-advert-slot`, as it looks to do a similar task to https://github.com/guardian/commercial/blob/main/src/lib/dfp/fill-advert-slots.ts which is filling slots with ads?

Reply:
Emphatic +1 from me, we can do better than `addSlot` (it's also hard to tell if someone is saying adSlot or addSlot when speaking aloud).

`fillDynamicAdSlot` sounds good!